xmtp_mls/src/groups/mod.rs

```diff
@@ -0,0 +1,304 @@
+                Some(group) => group,
+                None => return Err(GroupError::GroupNotFound),
+            };
+
+        Ok(mls_group)
+    }
+
+    pub fn create_and_insert(
+        client: &'c Client<ApiClient>,
+        membership_state: GroupMembershipState,
+    ) -> Result<Self, GroupError> {
+        let provider = client.mls_provider();
+        let mut conn = client.store.conn()?;
+        let mut mls_group = OpenMlsGroup::new(
+            &provider,
+            &client.identity.installation_keys,
+            &build_group_config(),
+            // TODO: Confirm I should be using the installation keys here
+            CredentialWithKey {
+                credential: client.identity.credential.clone(),
+                signature_key: client.identity.installation_keys.to_public_vec().into(),
+            },
+        )?;
+
+        mls_group.save(provider.key_store())?;
+        let group_id = mls_group.group_id().to_vec();
+        let stored_group = StoredGroup::new(group_id.clone(), now_ns(), membership_state);
+        stored_group.store(&mut conn)?;
+
+        Ok(Self::new(group_id, client))
+    }
+
+    pub async fn send_message(&self, message: &[u8]) -> Result<(), GroupError> {
+        let mut conn = self.client.store.conn()?;
+        let intent_data: Vec<u8> = SendMessageIntentData::new(message.to_vec()).into();
+        let intent =
+            NewGroupIntent::new(IntentKind::SendMessage, self.group_id.clone(), intent_data);
+        intent.store(&mut conn)?;
+
+        self.publish_intents(&mut conn).await?;
+        Ok(())
+    }
+
+    pub async fn add_members_by_installation_id(
+        &self,
+        installation_ids: Vec<Vec<u8>>,
+    ) -> Result<(), GroupError> {
+        let mut conn = self.client.store.conn()?;
+        let key_packages = self
+            .client
+            .get_key_packages_for_installation_ids(installation_ids)
+            .await?;
+        let intent_data: Vec<u8> = AddMembersIntentData::new(key_packages).into();
+        let intent =
+            NewGroupIntent::new(IntentKind::AddMembers, self.group_id.clone(), intent_data);
+        intent.store(&mut conn)?;
+
+        self.publish_intents(&mut conn).await?;
+
+        Ok(())
+    }
+
+    pub(crate) async fn publish_intents(&self, conn: &mut DbConnection) -> Result<(), GroupError> {
+        let provider = self.client.mls_provider();
+        let mut openmls_group = self.load_mls_group(&provider)?;
+
+        let intents = self.client.store.find_group_intents(
+            conn,
+            self.group_id.clone(),
+            Some(vec![IntentState::ToPublish]),
+            None,
+        )?;
+
+        for intent in intents {
+            // TODO: Wrap in a transaction once we can synchronize with the MLS Keystore
+            let result = self.get_publish_intent_data(&provider, &mut openmls_group, &intent);
+            match result {
```

Comment:
This is OK, but could remove some indentation 

```rust
if let Err(e) = result {
    log::error!(/* ... */);
    continue;
}

let (payload, post_commit_data) = result.expect("Checked for error; qed");
self.client.api_client.publish_to_group(vec![payload.as_slice()]).await?;
// ...
```

What happens if one of the intents errors & isn't published? I'm assuming we don't bubble the error b/c we need to ensure all intents are processed no matter what

Reply:
Great call-out. It's a bit of a tough situation. We don't want to have one bad intent completely brick your client because you can never move past it. On the other hand, we want the caller to be aware of the issue. I'm logging for now, but we probably should think about adding more state to keep track of the number of failed attempts or something.